### PR TITLE
introduce Reader::rc_records

### PR DESCRIPTION
**Pull request by TyberiusPrime**
This extracts and cleans up the RcRecords from #163. In brief, this is an iterator based API for BAM files that uses an Rc<> to avoid allocations. 

Speed in benchmarks is on par with the `read` based API ( see #163) - in my actual applications I have not been able to see a difference at all whatsoever.

This also introduces a 'sensible error handling' example for the `read` API, since I was working on those doc strings anyway.

### CODE REVIEW

File: src/bam/mod.rs [hunk 1]
```diff
@@ -92,14 +92,68 @@ pub trait Read: Sized {
...
+    /// Records iterator using an Rc to avoid allocating a Record each turn.
+    /// This is about 1% slower than the [`read`](#tymethod.read) based API in micro benchmarks,
+    /// but has nicer ergonomics (and might not actually be slower in your applications).
+    ///
+    /// Example:
+    /// ```
+    /// use rust_htslib::bam::{Error, Read, Reader, Record};
+    /// use rust_htslib::htslib; // for BAM_F*
+    /// let mut bam = Reader::from_path(&"test/test.bam").unwrap();
+    /// let mut record = Record::new();
```
Comment (tedil) [inline src/bam/mod.rs:136]:
```suggestion
```
I think the mutable record isn't needed here, right?

**Review by brainstorm — APPROVED** [review]

File: src/bam/mod.rs [hunk 2]
```diff
@@ -92,14 +92,71 @@ pub trait Read: Sized {
...
+    /// This is about 1% slower than the [`read`](#tymethod.read) based API in micro benchmarks,
+    /// but has nicer ergonomics (and might not actually be slower in your applications).
+    ///
+    /// Example:
+    /// ```
+    /// use rust_htslib::bam::{Error, Read, Reader, Record};
+    /// use rust_htslib::htslib; // for BAM_F*
+    /// let mut bam = Reader::from_path(&"test/test.bam").unwrap();
+    /// let record = Record::new();
+    //
```
Comment (tedil) [inline src/bam/mod.rs:137]: What I meant was that this `record` isn't used anywhere, so can be omitted completely :D
Aaaaand we lost a slash there

Comment (TyberiusPrime) [reply src/bam/mod.rs]: D'oh.

File: src/bam/mod.rs [hunk 3]
```diff
@@ -92,14 +92,70 @@ pub trait Read: Sized {
...
+    /// Records iterator using an Rc to avoid allocating a Record each turn.
+    /// This is about 1% slower than the [`read`](#tymethod.read) based API in micro benchmarks,
+    /// but has nicer ergonomics (and might not actually be slower in your applications).
+    ///
+    /// Example:
+    /// ```
+    /// use rust_htslib::bam::{Error, Read, Reader, Record};
+    /// use rust_htslib::htslib; // for BAM_F*
+    /// let mut bam = Reader::from_path(&"test/test.bam").unwrap();
+    //
```
Comment (tedil) [inline src/bam/mod.rs:136]:
```suggestion
    ///
```

**Review by tedil — APPROVED** [review]

**Pull request merged**